pkg/rbac/middleware.go

```diff
@@ -5,19 +5,42 @@ import (
 	"encoding/json"
 	"net/http"
 	"strings"
+	"sync"
+	"time"
 
 	"github.com/stolostron/search-v2-api/pkg/config"
 	authv1 "k8s.io/api/authentication/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/klog/v2"
 )
 
+//SAVING THIS IN ONLY BRANCH ANX/AUTHENTICATE
+
+//first we want to cache the authenticated tokenreview response:
+type UserTokenReviews struct {
+	uid             string
+	ValidatedTokens map[string]time.Time //token:timestamp
+	UserInfo        interface{}          //response from token review
+	expiresAt       time.Time
+	lock            sync.RWMutex // use this to avoid runtime error in Go fot simultaneously trying to read and write to a map
+
+	//future fields:
+	// In the future we'll add more data like
+	// hubClusterScopedRules []Rules
+	// rulesByNamespace      map[string]Rules
+	// managedClusters       []string // ManagedClusters visible to the user.
+}
+
 //verifies token (userid) with the TokenReview:
-func Middleware() func(http.Handler) http.Handler {
+func Middleware(utr *UserTokenReviews) func(http.Handler) http.Handler {
```

Comment:
`*UserTokenReviews` represents a single user.  Here need a structure that holds all users together. It could be an array, but later we may need something more.

Reply:
that makes sense - working on that change.